graylog2-server/src/main/java/org/graylog/plugins/views/search/searchtypes/pivot/buckets/ValuesBucketOrdering.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.plugins.views.search.searchtypes.pivot.buckets;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.graylog.plugins.views.search.searchtypes.pivot.BucketSpec;
+import org.graylog.plugins.views.search.searchtypes.pivot.PivotSort;
+import org.graylog.plugins.views.search.searchtypes.pivot.SortSpec;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class ValuesBucketOrdering {
+    private static boolean isGroupingSort(SortSpec sort) {
+        return PivotSort.Type.equals(sort.type());
+    }
+
+    private static boolean hasGroupingSort(List<SortSpec> sorts) {
+        return sorts.stream().anyMatch(ValuesBucketOrdering::isGroupingSort);
+    }
+
+    private static boolean needsReordering(List<? extends BucketSpec> bucketSpec, List<SortSpec> sorts) {
+        return bucketSpec.size() >= 2 && !sorts.isEmpty() && hasGroupingSort(sorts);
+    }
+
+    @VisibleForTesting
+    public static <T extends BucketSpec> List<T> orderBuckets(List<T> bucketSpec, List<SortSpec> sorts) {
+        if (!needsReordering(bucketSpec, sorts)) {
+            return bucketSpec;
+        }
+
+        final List<String> sortFields = sorts.stream()
+                .filter(ValuesBucketOrdering::isGroupingSort)
+                .map(SortSpec::field)
+                .collect(Collectors.toList());
+
+        return bucketSpec.stream()
+                .sorted((s1, s2) -> {
+                    int s1Index = sortFields.indexOf(s1.field());
+                    int s2Index = sortFields.indexOf(s2.field());
+                    if (s1Index == s2Index) {
+                        return 0;
+                    }
+                    if (s1Index == -1) {
+                        return 1;
+                    }
+                    if (s2Index == -1) {
+                        return -1;
+                    }
+                    return Integer.compare(s1Index, s2Index);
+                })
+                .collect(Collectors.toList());
+    }
+
+    public static Function<List<String>, List<String>> reorderKeysFunction(List<BucketSpec> bucketSpecs, List<SortSpec> sorts) {
```

Comment:
This method seems to be not unit-tested at all...

Reply:
Thanks, added tests!